javascript/ql/src/semmle/javascript/security/dataflow/ZipSlip.qll

```diff
@@ -0,0 +1,124 @@
+/**
+ * Provides a taint tracking configuration for reasoning about unsafe zip extraction.
+ */
+
+import javascript
+
+module ZipSlip {
+  /**
+   * A data flow source for unsafe zip extraction.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for unsafe zip extraction.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer guard for unsafe zip extraction.
+   */
+  abstract class SanitizerGuard extends TaintTracking::SanitizerGuardNode, DataFlow::ValueNode { }
+
+  /** A taint tracking configuration for Zip Slip */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "ZipSlip" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isSanitizerGuard(TaintTracking::SanitizerGuardNode nd) {
+      nd instanceof SanitizerGuard
+    }
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` in one step by virtue of
+   * `node2` being of the form `.pipe(node1)`. The reason this flow
+   * exists is that `.pipe` returns its argument to make chained
+   * stream operations work.
+   */
+  predicate pipeStep(DataFlow::Node node1, DataFlow::MethodCallNode node2) {
+      node2.getMethodName() = "pipe" and
+      node1 = node2.getArgument(0)
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` in one step including the assumption that
+   * `x` flows to `.pipe(x)`
+   */
+  predicate stepsThroughPipe(DataFlow::Node node1, DataFlow::Node node2) {
+    DataFlow::localFlowStep(node1, node2) or pipeStep(node1, node2)
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` including the assumption that
+   * `x` flows to `.pipe(x)`
+   */
+  predicate flowsThroughPipe(DataFlow::Node node1, DataFlow::Node node2) {
+    stepsThroughPipe*(node1, node2)
+  }
+
+  /**
+   * An access to the filepath of an entry of a zipfile being extracted
+   * by npm module `unzip`. For example, in
+   * ```javascript
+   * const unzip = require('unzip');
+   *
+   * fs.createReadStream('archive.zip')
+   *   .pipe(unzip.Parse())
+   *   .on('entry', entry => {
+   *      const path = entry.path;
+   *   });
+   * ```
+   * there is an `UnzipEntrySource` node corresponding to
+   * the expression `entry.path`.
+   */
+  class UnzipEntrySource extends Source {
+    UnzipEntrySource() {
+      exists(DataFlow::SourceNode parsed |
+        flowsThroughPipe(DataFlow::moduleImport("unzip").getAMemberCall("Parse"), parsed)
+        and
+        this = parsed.getAMemberCall("on").getCallback(1).getParameter(0).getAPropertyRead("path")
+      )
+    }
+  }
+
+  /**
+   * A sink that is the path that a createWriteStream gets created at.
```

Comment:
Suggestion: 

> The argument `arg` of `fs.createWriteStream(arg)`, as a sink for unsafe zip extraction.

We usually have "..., as a X for Y" for these simple source/sink/sanitizer boxes.

The explanation of why we need this sink can go into a non-docstring comment.

Reply:
Done. Am I formatting the non-docstring comments according to local convention? I checked https://github.com/Semmle/ql/blob/master/docs/ql-style-guide.md and didn't see any specific preference between `//` and `/* */` but I don't see any `/* */` nearby in the code.

Reply:
I think this is just an opinion of the purpose of `//` and `/**` comments, it would apply equally to java code.

The explanation of why we need this sink is tied to a particular implementation elsewhere, something that a *user* of this class does not need to care about, but a QL *developer* may wonder about that.

The `/**`-comments become the docstring of the class/predicate/module, and will be shown on pages like this one: https://help.semmle.com/qldoc/javascript/semmle/javascript/dataflow/AbstractValues.qll/type.AbstractValues$AbstractArguments.html.

The `//` are meant for internal consumption.

Reply:
I was merely talking about the difference between preferences for single line vs. multi-line in *non*-documentation comments. The intentions around documentation vs. non-documentation comments are fairly clear.

Reply:
In that case, it seems that single line comments dominate for JavaScript, but I do not think there is an official policy:

```
$ ag '/\* ' -G qll\?\$ ql/javascript
ql/src/semmle/javascript/dependencies/FrameworkLibraries.qll
371:  override string getAnEntryPoint() { /* not easily detectable */ none() }
414:  override string getAnEntryPoint() { /* not easily detectable */ none() }
453:  override string getAnEntryPoint() { /* not easily detectable */ none() }

ql/test/tutorials/Validating RAML-based APIs/RAML.qll
1:/* Model of RAML specifications. */

ql/test/tutorials/Validating RAML-based APIs/Osprey.qll
1:/* Model of Osprey API implementations. */
```

But the other languages appear to use another style:

```
ag '/\* ' -G qll\?\$ ql | wc -l
641
```

Reply:
Those results match my searching around, yeah.